### PR TITLE
Nub 895 working example of collections API

### PR DESCRIPTION
**Pull request by macsrok**

### CODE REVIEW

File: lib/cloudapp/cli/collections_prompt.rb [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            list_users collection_id access_id
+                - remove a user from the collection referenced by collection_id. access_id is required.
+
+            help
+                - print this message
+            
+EOS
+      end
+
+      def self.parse_options(args)
```
Comment (joiggama) [inline lib/cloudapp/cli/collections_prompt.rb]: Hey @macsrok what do you think about using  ruby's OptParser for collections subcommand so we avoid the case statement?

File: cloudapp.gemspec [hunk 2]
```diff
@@ -43,17 +43,20 @@ Gem::Specification.new do |s|
...
-  s.add_dependency 'netrc',      '~> 0.7.7'
-  s.add_dependency 'typhoeus',   '~> 0.7.3'
+  s.add_dependency "clipboard", "~> 1.0.1"
+  s.add_dependency "leadlight", "~> 0.1.0"
+  s.add_dependency "mime-types", "~> 1.19"
+  s.add_dependency "netrc", "~> 0.7.7"
+  s.add_dependency "typhoeus", "~> 0.7.3"
+  s.add_dependency "http", "~> 0.7.4"
+  s.add_dependency "terminal-table"
+  s.add_dependency "pry"
```
Comment (joiggama) [inline cloudapp.gemspec]: @macsrok `pry` should not be a runtime dependency, can you move it to development deps ?

File: lib/cloudapp/cli/collections.rb [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    class Collections
+      COLLECTION_FIELDS = ["id", "name", "shared", "is_public", "created_at", "updated_at"]
+      ITEM_FIELDS = ["id", "slug", "name", "item_type", "view_counter", "favorite", "share_url", "created_at", "updated_at"]
+
+      def initialize(token)
+        @BASE_URL = "https://my.cl.ly/v4/collections"
+        @HTTP = HTTP.auth("Bearer #{token}")
+          .with(accept: "application/json", content_type: "application/json")
+          .accept(:json)
+      end
```
Comment (joiggama) [inline lib/cloudapp/cli/collections.rb]: @macsrok you should probably down-case instance variable names. Also, adding an `attr_reader` for `http`, would help you access directly without `@`.

File: lib/cloudapp/cli/collections.rb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+      def self.token_to_jwt(token)
```
Comment (joiggama) [inline lib/cloudapp/cli/collections.rb]: I'd recommend moving class methods top the top.

File: lib/cloudapp/cli/collections.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+          h.select! { |k, v| COLLECTION_FIELDS.include? k }
+          collections << h.values
+        end
+
+        table = Terminal::Table.new :title => "Listing Collections", :headings => ["Collection ID", "Collection Name", "Shared", "Public", "Created At", "Updated At"], :rows => collections
+        puts table
+      end
+
+      def list_items(collection_id)
+        resp = JSON.parse(@HTTP.get("#{@BASE_URL}/#{collection_id}").body.readpartial)
```
Comment (joiggama) [inline lib/cloudapp/cli/collections.rb]: @macsrok It'd be useful to have a url helper method, so you can only pass a route path and get back the full URL to dry this up a little bit:

```ruby
def url_for(path = "")
  "#{@base_url}/#{path}"
end
```

What do you think?

Comment (joiggama) [reply lib/cloudapp/cli/collections.rb]: Then you could do:

```ruby
def list_items(collection_id)
  http.get(url_for(collection_id))
  # ...
end
```

File: lib/cloudapp/cli/collections.rb [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        items = []
+        resp["items"].each do |i|
```
Comment (joiggama) [inline lib/cloudapp/cli/collections.rb]: What do you think about:

```ruby
items = resp["items"].map{ |i| i.slice(*ITEM_FIELDS).values }

File: lib/cloudapp/cli/collections.rb [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+          i.select! { |k, v| ITEM_FIELDS.include? k }
+          items << i.values
+        end
+        table = Terminal::Table.new :title => "Listing Items in Collection `#{resp["name"]}`", :headings => ["Item Slug", "Item Name", "Created At", "Updated At", "Item Type", "Item View Count", "Item Share URL", "Item ID", "Item Favorited"], :rows => items
```
Comment (joiggama) [inline lib/cloudapp/cli/collections.rb]: Favor newer hash syntax, also, for readability, can you split this line into multiple ?

Comment (macsrok) [reply lib/cloudapp/cli/collections.rb]: resolved with commit 4b8e602b06153cd6a8d1258b292dedc4c78a43b0

File: lib/cloudapp/cli/collections.rb (hunk 4 above)
Comment (macsrok) [reply lib/cloudapp/cli/collections.rb]: resolved with commit 4b8e602b06153cd6a8d1258b292dedc4c78a43b0

File: lib/cloudapp/cli/collections.rb (hunk 6 above)
Comment (macsrok) [reply lib/cloudapp/cli/collections.rb]: Rails is required for Hash#slice

Comment (joiggama) [reply lib/cloudapp/cli/collections.rb]: Ahh, you're right... active support hurts our brains.. 😧

**Comment by joiggama** [comment]
@macsrok have you been able to run existing specs, do you know if we're in good shape to add some test scenarios for collections ?

**Pull request merged**